lib/src/license_detection/license_detector.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:meta/meta.dart';
+import 'package:pana/src/third_party/diff_match_patch/diff.dart';
+import 'package:source_span/source_span.dart';
+import 'package:string_scanner/string_scanner.dart';
+
+part 'confidence.dart';
+part 'crc32.dart';
+part 'license.dart';
+part 'primary_filter.dart';
+part 'token_matcher.dart';
+part 'tokenizer.dart';
+
+// Load corpus licenses.
+final _licenses = loadLicensesFromDirectories(_directories);
+
+// WIP: Returns a list of detected licenses whose
+// confidence score is above a certain threshold.
+List<LicenseMatch> detectLicense(String text, double threshold) {
+  final granularity = computeGranularity(threshold);
+
+  final unknownLicense =
+      LicenseWithNGrams.parse(License.parse('', text), granularity);
+
+  final possibleLicenses = filter(unknownLicense.occurrences, _licenses)
+      .map((e) => LicenseWithNGrams.parse(e, granularity));
+  var result = <LicenseMatch>[];
+
+  for (var license in possibleLicenses) {
+    final matches = findPotentialMatches(
+      unknownLicense,
+      license,
+      threshold,
+    );
+
+    for (var match in matches) {
+      try {
+        final hit = licenseMatch(unknownLicense, license, match, threshold);
+        result.add(hit);
+      } on LicenseMismatchException catch (_) {
```

Comment:
We were going with that approach at first but then Jonas it would be better to throw a mismatch exception and we can avoid checking null condition.

Reply:
Returning an object with zero score might be an okay way to do it.

I figured the exception is better than null, because we don't get nullablility in everywhere.